RFC0008 - Module system (#101)
Draco-lang/Language-suggestions: Specification/ModuleSystem.md

# Module System
This RFC defines how code is split to multiple files, how symbols are exported and imported and how reusable units of code are defined.  
## Packages and modules

> A package is a reusable unit of code, which consists of modules. In C# terminology, it is equivalent to a project that can be distributed as a NuGet package. Module is portion of code equivalent to draco file (such modules will be called just modules from now), or a directory in draco project (such modules will be called module folder). Module name is same as name of draco file without file extension, or as folder name. Modules can also be defined in code using the `module` keyword followed by name of the module and code block. Modules declared like this will become submodule of the module they were declared in. Any draco file or a folder declared under module folder will become submodule of the module folder. Module folders can group their contribution under the folder's module name by having file named `module.draco` under the module folder. Any code declared in this file, will be visible in module of its module folder. In code names of submodules are separated by `.`. Modules are not visible by default, they need to be exported by their parent module to become visible and then they become part of the public api of the parent module. Project name will become root module for given project. This module will not have any code in it. Any modules in the root folder of the project will be visible by default.  

Reviewer: > In C# terminology, it is equivalent to a project that can be distributed as a NuGet package

You mention both packages and modules in the previous sentence, it's unclear which one you refer to. You could either state it explicitly or talk about modules later.

Reviewer: > Module name is same as name of draco file without file extension, or as folder name.

Here a single example in a sentence would be great. "For example, `math.draco` defines the module named `math`"

Reviewer: I'd might tackle the issue of modules in a completely separate section and try to approach it from a hierarchy perspective. Our modules make up a tree hierarchy, it might be easier to explain terminology and elements through that. (we could discuss on the server exactly how)

## Exporting symbols

> Each module is responsible for exporting their own API towards other modules. Any symbol that should be accessible from the outside must be marked for exporting, using the `export` keyword. Symbols can be exported inline, by adding the `export` keyword before the symbol declaration or as export list, allowning us to alias symbols at the export. The syntax for exporting list of symbols is `export <symbol list>`, where `<symbol list>` can be ither symbol name with optional alis using the `as` keyword with the syntax `<symbol name> as <new name>`, or as list of such symbols encapsulated in curly braces and separated by comma.  

Reviewer: > Each module is responsible for exporting their own API towards other modules.

This is a key observation, really nice!

Example of exporting:
```js
// Exporting members in an export list, aliasing y
export { y as second, add1 };

// Exporting x inline
export var x = 0;

var y: int32;

func add1(x: int32): int32 = x + 1;
```
Symbols can also be re-exported using the `from` keyword with syntax `from <module name> export <symbol list>` where instead of symbol list we can also use `*` to export everything.  
Example of re-exporting:
```js
// Assuming this is foo.fr

// Now, everything is accessible from foo, that was exported by bar, as is
from bar export *;

// Aliased add1, expose everything else as-is from baz
from baz export { add1 as add_one, * };
```
Syntax for exporting modules is `export <module name>`.  
**Note**: the module is only visible by its parent module before export, meaning that this is only valid at the module parent.
## Imporing symbols

> Symbols can be accessed through their fully qualified name, but importing into local scope can be done with the `import` statement, that has syntax `from <module name> import <symbol list>` where symbol list has the same syntax as as symbol list for exporting, meaning that symbols can also be aliased.

Reviewer: > Symbols can be accessed through their fully qualified name

That's not 100% correct, they can only be referenced through the qualified name if the module path has been fully exported down the line, meaning that in `x::y::z::foo()`, `foo` was exported in `z`, `z` was exported in `y`, `y` was exported in `x` and `x` was exported in the top-level module.

# Module System

> This RFC defines how we split up code to multiple files, export symbols, import symbols and how do we define reusable units of code.  

Reviewer: ```suggestion
This RFC defines how we split up code to multiple files, export symbols, import symbols and how we define reusable units of code.  
```

Reviewer: Some rephrasing should be done to exclude 'we' from the picture. This is specification text.

## Packages and modules

> Package is reusable unit of code, which contains one or more modules, it is equivalent to a c# project and we can generate one nuget package from it.  

Reviewer: ```suggestion
A package is a reusable unit of code, which consists of modules. In C# terminology, it is equivalent to a project that can be distributed as a NuGet package.
```

> Module is portion of code. Module hierarchy follows the directory structure, meaning that file `bar.draco` in directory `foo` produces module `foo.bar`. Module folders can group their contribution under the folder's module name. For example, if a folder `math` has files `marix.draco`, `vector.draco` and `quat.draco`, a file called `module.draco` inside `math` can provide a single module interface for all of these. Modules are not visible by default, they need to be exported by their parent module to become visible and then they become part of the public api of the parent module.  

Reviewer: You are phrasing it extremely densely, it's hard to grasp.

Reviewer: An example is nice, but introduce the conceptual words like package, module, module hierarchy more precisely. The example can come after, and definitely graph out the directory structure in a code block, not just in text.

## Exporting symbols

> Each module is responsible for exporting their own API towards other modules. Any symbol that should be accessible from the outside should be marked for exporting, using the `export` keyword. Symbols can be exported inline, by adding the `export` keyword before the symbol declaration or as export list, allowning us to alias symbols at the export with the syntax `export <symbol list>`, where `<symbol list>` can be ither symbol name with optional alis using the `as` keyword with the syntax `<symbol name> as <new name>`, or as list of such symbols encapsulated in curly braces and separated by comma.  

Reviewer: This sentence spans a whopping 3 lines! While it's not particularly a complex one, it's definitely hard to understand as a whole, even on 2nd or 3rd read. I suggest breaking it up.

Example of exporting:
```js
// Exporting members in an export list, aliasing y
export { y as second, add1 };

// Exporting x inline
export var x = 0;

var y: int32;

func add1(x: int32): int32 = x + 1;
```

> Symbols can also be reexported using the `from` keyword with syntax `from <module name> export <symbol list>` where instead of symbol list we can also use `*` to export everything.  

Reviewer: ```suggestion
Symbols can also be re-exported using the `from` keyword with syntax `from <module name> export <symbol list>` where instead of symbol list we can also use `*` to export everything.  
```

> Example of reexporting:

Reviewer: ```suggestion
Example of re-exporting:
```

# Module System
This RFC defines how code is split to multiple files, how symbols are exported and imported and how reusable units of code are defined.  
## Packages and modules

> A package is a reusable unit of code, which consists of modules. In C# terminology, package is equivalent to a project that can be distributed as a NuGet package. Module is portion of code equivalent to draco file (such modules will be called just modules from now), or a directory in draco project (such modules will be called module folder). Module name is same as name of draco file without file extension, or as folder name. For example for file `foo.draco` the module name would be `foo` and for folder named `bar` the module name would be also `bar`. Modules can also be defined in code using the `module` keyword followed by name of the module and code block with the contents of the module. Modules declared like this will become submodule of the module they were declared in. Any draco file or a folder declared under module folder will become submodule of the module folder. Module folders can group their contribution under the folder's module name by having file named `module.draco` under the module folder. Any code declared in this file, will be visible in module of its module folder. Contents of module (submodules are also contents of module) can be accessed with the syntax `<module name>.<name of symbol to access>`. All modules are visible in the package they were declared in, if they should be visible in other package that imports the package this modules were declared in, they need to be exported by their parent module to become visible and then they become part of the public api of the parent module. Project name will become root module for given project. This module will not have any code in it. For modules to be visible from outside package they were declared in, they need to be exported from the root of the package.  

Reviewer: > Module is portion of code equivalent to draco file

This is imprecise, a file can actually define multiple modules by the issue currently (yes you expand on this later but we shouldn't start the specs with an imprecision like this). Modules should be explained as the building blocks of packages, that provide or hide pieces functionality to make up the package or something.

Reviewer: > Module is portion

> Module name

I'm missing a couple of articles all around. When in singular, it reads more natural to start with an article. "A module", "The module", ...

Reviewer: > Modules declared like this will become submodule of the module they were declared in

Yes, but the modules declared within the directory are also submodules. I know you expand on this later, but this is why I suggested approaching all of this from a standpoint of hierarchy. You are duplicating a lot of knowledge between nesting modules within a file and nesting within folders.

## Exporting symbols
Each module is responsible for exporting their own API towards other modules. Any symbol that should be accessible from the outside must be marked for exporting, using the `export` keyword. Symbols can be exported inline, by adding the `export` keyword before the symbol declaration or as export list, allowning us to alias symbols at the export. The syntax for exporting list of symbols is `export <symbol list>`, where `<symbol list>` can be ither symbol name with optional alis using the `as` keyword with the syntax `<symbol name> as <new name>`, or as list of such symbols encapsulated in curly braces and separated by comma.  
Example of exporting:
```js
// Exporting members in an export list, aliasing y
export { y as second, add1 };

// Exporting x inline
export var x = 0;

var y: int32;

func add1(x: int32): int32 = x + 1;
```

> Symbols can also be re-exported using the `from` keyword with syntax `from <module name> export <symbol list>` where instead of symbol list we can also use `*` to export everything.  

Reviewer: I'm missing an explanation on what re-exporting actually is.

Example of re-exporting:
```js
// Assuming this is foo.fr

// Now, everything is accessible from foo, that was exported by bar, as is
from bar export *;

// Aliased add1, expose everything else as-is from baz
from baz export { add1 as add_one, * };

> ```

Reviewer: Let's come up with a more practical example for a specs. The one later is really good!

Syntax for exporting modules is `export <module name>`.  

> **Note**: the module is only visible by its parent module before export, meaning that this is only valid at the module parent.

Reviewer: This sentence is highly ambiguous and hints that it won't be visible after export.

# Module System
This RFC defines how code is split to multiple files, how symbols are exported and imported and how reusable units of code are defined.  
## Packages and modules

> A package is a reusable unit of code, which consists of modules. In C# terminology, package is equivalent to a project that can be distributed as a NuGet package. Modules are building blocks of packages, they provide or hide pieces functionality that makes up the package, module can be draco file (such modules will be called just modules from now), or a directory in draco project (such modules will be called module folder), or it can be defined in code using the syntax `module <module name> <code block>`. Module name is same as name of draco file without file extension, or as folder name, or if the module is defined in code, the module has the name that was specified for it in code. For example for file `foo.draco` the module name would be `foo` and for folder named `bar` the module name would be also `bar`. Any draco file or a folder declared under module folder will become submodule of the module folder. Modules defined in code will become submodules of the module they were declared in. Module folders can group their contribution under the folder's module name by having file named `module.draco` under the module folder. Any code declared in this file, will be visible in module of its module folder. Contents of module (submodules are also contents of module) can be accessed with the syntax `<module name>.<name of symbol to access>`. All modules are visible in their parent module, if they should be visible elsewhere, they need to be exported by their parent module to become visible and then they become part of the public api of the parent module. Project name will become root module for given project. This module will not have any code in it. For modules to be visible from outside package they were declared in, they need to be exported from the root of the package.  

Reviewer 2: Break the line more often, this line fill more than my gigantic smartphone screen.
Adding space make text more readable too.  

About projects:
A nuget package can contains multiples DLLs
A DLL can be the product of multiples projects (source imports is common in some corners of .NET).

# Module System
This RFC defines how code is split to multiple files, how symbols are exported and imported and how reusable units of code are defined.  
## Packages and modules

> A package is a reusable unit of code, which consists of modules. In C# terminology, package is equivalent to a project. Modules are building blocks of packages, they provide or hide pieces functionality that makes up the package. Module can be draco file, the module name of such file being name of this file without file extension. Module can also be a directory in draco project, (such module will be called module folder). For example for file `foo.draco` the module name would be `foo` and for folder named `bar` the module name would be also `bar`. Module can also be defined in code using the syntax `module <module name> <code block>`.   

Reviewer 2: Same comment than on the other specs, the text is too dense, give more room for the reader to breath, try to start sentences on a new line more often.  
At least here the sentence are shorts enough, you only need to change the spacing.

Reviewer 2: > A package is a reusable unit of code

Doesn't mean a lot IMO, and may also be wrong. Does a console app be a package ? If yes, the code here is not intended to be reusable.

Reviewer 2: > they provide or hide

replace `provide` with `expose`

Example of exporting:
```js
// Exporting members in an export list, aliasing pow
export { pow as power, abs };

// Exporting PI inline
export val PI = 3.1415;

func pow(num: int32, exp: it32): int32{
  var i = 0;
  var result = 1;
  while(i < exp){
    result = result * num;
    i = i + 1;
  }
  return result;
}

func abs(x: int32): int32 = if(x > 0) x else -x;
```
Symbols can also be re-exported using the `from` keyword with syntax `from <module name> export <symbol list>` where instead of symbol name we can also use `*` to export everything. Re-exported symbols will be propagated to the parent module. Meaning that the parent module will contain the symbols directly.  
Example of re-exporting:
```js
// Assuming this is module math

// Now, everything is accessible from math directly

> from constants export { GravityConstant as G, * };

Reviewer 2: I have an issue with that, there is a lot of ambiguity about what it is happening here.  
For example, is GravityConstant exported two time, if not, why ?
This specs here doesn't specify this behavior too.
